### ISSUE TITLE
OpenGL: Pinch zoom to magnify text / icons

### ISSUE THREAD

**Issue by sonora** [osmandapp/OsmAnd#15034]
It may not be easy to fix, but let me trigger the investgation anyway: Using legacy rendering, if in bad light conditions like direct sunlight, and trying to read text on the map, you can use "pinch zoom and hold", i.e. pinch-zooming the map and leaving your fingers in position, to temporarily magnify a map area (including all text) without it getting re-rendered, this e.g. facititates reading small or tilted map text.

Using OpenGL rendering, this trick fails, because map text immediately disappears during the maneuver, thus is never magnified.

Any chance to fix OpenGL to support that trick again?

**Comment by vshcherb** [comment]
This wouldn't work with Intermediate zoom feature which we plan to fix. So text should be increased only with a setting of Text size.

**Comment by vshcherb** [comment]
#15009

**Comment by sonora** [comment]
Let me perhaps reopen this to see if we can find a solution.

This issue is about (with the OpenGL engine) being able to use pinch zoom as an on-the-fly  screen magnifier, like the V1 engine allows.

We had interim linked it to #15009, but unfortunately the fix for that did not solve this issue here.

**Comment by vshcherb** [comment]
I still don't think we should change anything, I don't think it's important feature but let's keep it open

**Comment by Gregor0000** [comment]
I think this feature is quite important and it is one of the major reasons, I don't use OpenGL rendering yet.
It is so convenient to be able to increase the text size with pinch&zoom in situations where the readability of the screen is bad (bad light - too dark, too bright, when wearing sunglasses). For me, those situations happen all the time during my hikes. Increasing the overall text size is not a good solution: It clutters the map view and makes a good overview more difficult.
I would very much appreciate if such a functionality could be implemented.

**Comment by sonora** [comment]
I wonder if the feature requested in this issue could not be achieved by wrapping the content of this method https://github.com/osmandapp/OsmAnd-core/blob/a37dfcced5fb0ce9f0783ee3707ea7aa6cf25545/wrappers/android/src/net/osmand/core/android/MapRendererView.java#L1734 in a condition like
```
public void onDrawFrame(GL10 gl) {
	if (!(OsmandMapTileView).isPinchZoomingOrRotating()) {

		(...)

	}
}
```
which would halt any re-rendering until the fingers are taken off the screen, just like the legacy renderer handles things...

**Comment by vshcherb** [comment]
It will break zoom out functionality as there it's needed to have rendering, I'm not sure it's a feature, it was a bug in legacy rendering cause it wasn't possible not to scale. Though for some users it turns to be a feature however we have text size

**Comment by sonora** [comment]
Yes, could certainly limit to apply only for zoom-in cases.

In my experience it's not really only about text, it's all sorts of stuff sometimes hard to see in e.g. bright sunlight.

I personally sometimes use for contour lines, where (and rightly so) we use small and thin font for the elevation labeling which you do not read very often. But in some situations you need to - a perfect use case for such a quick "temporary magnification". Or to quickly inspect a highway shield label you would also not like to permanently show overly large.

Workaround (but much clumsier) could be an explicit temporary screen-magnifier via special quick action button. But using pinch for this is just REALLY intuitive and convenient.  :wink:

**Comment by sonora** [comment]
An alternative strategy could be this:

1. Capture the scale factor from the gesture detector during the pinch-zoom gesture (`Matrix.scaleM(textMatrix, 0, scaleFactorFromPinch, scaleFactorFromPinch, 1.0f);`)
2. When rendering text, apply the captured scale factor to the text’s transformation matrix before drawing it (`renderTextWithMatrix(textMatrix);`).
3. After the pinch-zoom gesture ends, update the text size based on the final scale factor and re-render the text at its new size.

**Comment by vshcherb** [comment]
Too much for me.... @sonora can you implement flag that will be enabled in your build and code could be shared. So you will continue to have own version with it.

Good place to have such code in Version class where we have flags for different builds (paid non paid ,...)

**Comment by sonora** [comment]
Will be happy to experiment with this, need to find some major time, I guess. :wink: The simple goal is to suggest the least intrusive solution which could be turned on/off via a single flag.

 Major challenge for me currently is how in a class like e.g. `MapRendererView.java` belonging to **osmand-core** I can import something or create awareness of a gesture state which is e.g. handled here `(OsmandMapTileView).isPinchZoomingOrRotating()` in **OsmAnd**.  Any hint for me?

**Comment by vshcherb** [comment]
Well you already maintain specific changes, the goal that you don't maintain specific code and maintain only 1 line of specific code in Version. All the rest could should be merged to master, later it could be specific Flavor build. We do very similar already to different market versions but we don't maintain separate code

**Comment by sonora** [comment]
Sure we can do that if you like! But at this point I was asking about our coding hygiene: I notice that classes we have in OsmAnd-core do never import any classes from OsmAnd, or am I worong? So what's the procedure of choice to share methods or context between these?

**Comment by vshcherb** [comment]
Of course core is completely independent library and couldn't use any classes from OsmAnd, so in order to pass code https://en.wikipedia.org/wiki/Inversion_of_control should be used

**Comment by sonora** [comment]
Perhaps I am mistaken, but it seems [this commit](https://github.com/osmandapp/OsmAnd-core/commit/4b8e57d2d50c0c722c213e731c60e12402da3739)  by @Andrewschenko may lay the foundation for a feature where we 
- suspend symbol updates during pinch-zooming,
- temporarily bitmap-scale them,
- and re-render them only after the fingers are removed from the screen?

Roughly like this:

1. Extend the IMapRenderer to include gesture-related hooks:
```
virtual void onPinchZoomStart() = 0;
virtual void onPinchZoomUpdate(float scale) = 0;
virtual void onPinchZoomEnd() = 0;
```

2. Modify the renderer / MapRenderer.cpp
Add a flag like bool isPinchZooming = false;
Add a scale factor like float pinchZoomScale = 1.0f;
```
if (isPinchZooming) {
    applyTemporaryLabelScale(pinchZoomScale);
    drawLabels(); // Use cached labels, scaled
} else {
    drawLabels(); // Normal rendering
}
```

3.  Suspend symbol updates during gesture:
```
suspendSymbolsUpdate(); // On pinch start
resumeSymbolsUpdate();  // On pinch end

```

4. Apply Scale in OpenGL
```
glPushMatrix();
glScalef(pinchZoomScale, pinchZoomScale, 1.0f);
drawLabels();
glPopMatrix();
```

5. Trigger re-render after gesture ends:
`forcedFrameInvalidate(); // Forces full re-render with updated label layout
`

**Comment by isso-git** [comment]
I fully support this suggestion, if it´s possible in any way it should be done.
I am often switch to rendering engine version 1 to make use of this feature.
It's these small, useful details that make software stand out from the rest.

**Comment by sonora** [comment]
@RZR-UA @Andrewschenko @alex-dev-neo @ihoryakovenko @ivanPyrohivskyi 

With all the great and successful work being done on the OpenGL rendering lately, is it possible we take a look at this 'transient text magnification' functionality again (like our V1 engine has it)?

**Comment by vshcherb** [comment]
Will get back to it

**Comment by vshcherb** [comment]
This could be implemented as Accessibility setting cause no other map applications do it. So if you can prepare pull request for it, it could be ok @sonora but I see it as a multi step process. Let's investigate it first how much work to be done and do it step by step @Andrewschenko

**Comment by vshcherb** [comment]
Please note that there is another task to magnify text and icons by separate settings

**Comment by sonora** [comment]
@vshcherb , @Andrewschenko I had done some research and talked to various people how this would be properly implemented in OpenGL, and had captured my notes over time in the various post above. I do, however, not feel confident to attempt a founded implementation in our core code, would feel much better if one of our seasoned devs could take a look. :wink:

Regarding "Accessibility": Yes, this could certainly be a switch / feature in the Accessibility section to activate/deactivate, if helpful. There are also other ways of implementing it, like a 'temporary screen magnification feature' upon some special gesture. **But the beauty about how we have it in the V1 engine** is that it does not interfere at all with 'regular' zooming and map handling. It never needed explaining or documentation either. There it is 'always available', by simply leaving your fingers on after pinch-zooming in, no separate setting or switch needed, and if you never used it no problem or complication either...

**Comment by Andrewschenko** [comment]
It looks like it could be implemented as some optional feature (switchable somewhere), so the simple two-finger pinch to zoom would simply rescale the viewport instead - just to make it look bigger on the screen (while fingers are still touching the screen).

**Comment by DmitryAlexei** [comment]
- How should this tool work when zooming out?

**Comment by sonora** [comment]
> * How should this tool work when zooming out?

For zooming out this has no functional benefit and can simply behave exactly like today when you try the V1 engine. (I.e. simply scaling the bitmap down until finger release. It's unnoticed by most, because no one usually keeps the fingers on the screen after zooming _out_.)

**Comment by ihoryakovenko** [comment]
- Do we need to support this feature for Android Auto?

**Comment by vshcherb** [comment]
Nope AA is not needed

**Comment by sonora** [comment]
True, not needed. Mainly essential for outdoor use, bright light conditions, lots of screen glare and reflections, etc.

And PS, regarding a switch to turn this on or off: We could of course have it, but perhaps not needed:
- Consider that we have always had this in V1, anyway.
- And consider that this anyway only kicks in if you zoom _and leave your finger on the screen at the end of the gesture_. A highly unusual behavior, unless you intentionally go for this transient magnification effect, actually.

**Comment by ihoryakovenko** [comment]
- [x] Add switch to Accessibility plugin

**Comment by sonora** [comment]
> * [ ]  Add switch to Accessibility plugin

@vshcherb We are still early in the 5.2 release cycle, and as I said [above](https://github.com/osmandapp/OsmAnd/issues/15034#issuecomment-3250035664), this is non-interfering with normal user behavior. So maybe we can try just silently activating by default, and see if there is any adverse effect or user feedback?

**Comment by ihoryakovenko** [comment]
QA: Enable accessibility plugin, set - Enable pinch zoom magnification. Check Pinch zoom on devices with different screen resolutions, as this option is hardware-dependent, and the max zoom can be different.

**Comment by sonora** [comment]
First tests: It's great!

3 glitches from what I see, try how it behaves differently vs. V1 still in these respects:
1. The temporary magnification is not centered at the _zoom_ center (middle point between the 2 fingers), but at the _screen_ center. Compare how V1 works different in this respect .(Not always is what you need to enlarge in the screen center.)
2. Map panning becomes disabled now during the pinch-zoom gesture when zooming in (still works while pinch-zooming out). In V1 the map always pans along if the fingers shift sideways during the pinch gesture. (Perhaps this is connected with point 1 above).
3. The magnification effect seems to work only for 2 finger pinch zooming, but not yet for the one finger (double-tap plus slide) zoom gesture.

**Comment by ihoryakovenko** [comment]
- [x] Zoom center should be between 2 fingers, not always in the screen center.
- [x] Switch in Accessibility plugin should not be blocked behind Android accessibility mode

**Comment by sonora** [comment]
@ihoryakovenko  Better, THANKS!

Small glitch still regarding the panning: Magnify something near the edge of the screen and try to drag it to the center: The map will move the **opposite** way than expected (again compare vs. how V1 behaves).

**Comment by sonora** [comment]
Remaining things to consider:

- [ ] Dragging (panning) works in opposite direction than in V1. (Magnify something near the edge of the screen and try to drag it to the center.)
- [ ] In V1 this is "always on". In V2 we made a switch in the Accessibility plugin.
  - Is the switch needed, would it impact anybody to simply make it also "always on"?
- [ ] In V1 this works also for the one finger double-tap plus slide gesture. Shall we activate it for V2 likewise?

**Comment by ihoryakovenko** [comment]
@sonora  Dragging (panning) works in opposite direction than in V1 is expected behavior if we use viewport rescale. New tiles are not loaded, so I cannot move the camera over the border of already loaded tiles.
To achieve V1 behavior, instead of moving the offset, the viewport rescale should affect the getLocationFromScreenPoint function. And I don't know if it is possible @Andrewschenko

**Comment by sonora** [comment]
Got it, @ihoryakovenko , thanks for explaining!

But (also @Andrewschenko ):  This leads to this curious effect:: Use the V2 engine with the new feature activated:

- Now, instead of pinching to zoom **_in_**, pinch to zoom **_out_** and shift the fingers a bit: The panning will work as expected (like in V1).
- But as soon as you reverse the pinch to now zoom in (even if still within the same gesture), the panning action will reverse.

To be sure: It's **much** better than we had before without this feature. But it s a weird and inconsistent behavior...  :wink:

**Comment by ihoryakovenko** [comment]
https://github.com/user-attachments/assets/e1d628d5-2e42-41b7-a973-9ca37e7b15ef

**Comment by ihoryakovenko** [comment]
@sonora, I found a solution for map movement. Other things to consider are up to the PO.

**Comment by sonora** [comment]
**_Fantastic_**, @ihoryakovenko  !!  I just did a brief check and it works fabulously! (2D as well as 3D!)

MANY THANKS for your work here!!

**Comment by sonora** [comment]
@ihoryakovenko 
- [ ] Panning now overshoots the pinch center. When you slightly pinch-zoom in, then pan a little, the map moves more than the panning, i.e. the zoom center (middle between the fingers) does not stay at the same point on the map. It's hard to see at far-out zooms, but zoom in towards the maxzoom and try then, there it's very obvious.
- [ ] There's also an initial "screen jump", noticeable at high zoom level.

EDIT: This patch https://github.com/osmandapp/OsmAnd/pull/23474 fixes:
- the overshoot issue (by re-centering to the initial tile)
- an issue with initial "screen jumps" (most noticeable at high zoom and in 3D) coming from using different coordinate systems in `onZoomStarted(...)` vs. `onChangeViewAngleStarted()`

**Comment by sonora** [comment]
@DmitryAlexei @ihoryakovenko I have merged the small patch so we can test directly with it.

My testing:
1. It seems we have a slight "screen jump" at the end of a pinch zoom gesture between the magnified viewport and the then re-rendered viewport. Possibly a small tile-misalignment, similar to the one we had at the beginning of the gesture and fixed by my patch.
2. There are spurious bigger jumps I believe occur if you start a pinch zoom gesture very early before all tiles in the view port are determined/rendered yet. _EDIT: This also occurs with pinch zoom magnification disabled, i.e. is a weakness of V2, not the new feature._

Both effects deserve investigation, perhaps mitigation, but have rather little impact on normal app use.

**Comment by Zirochkabila** [comment]
# Testing 1

`OsmAnd~ 5.2.0#6051m, Released: 2025-09-15`

The following has been verified (comparing V1 & V2):
1. Zoom is centered on the zoom scale (screen)
2. Map panning occurs during pinch-to-zoom gesture
3. Zoom effect is present when double-tapping + moving your finger
4. Panning + dragging works in the desired direction


<details><summary>Details</summary>
<table style="width:100%">
  <tr>
    <td><video source src="https://github.com/user-attachments/assets/52dac381-192c-4736-bab4-51c1e02714a2"  width="300" height="auto"/></td>
    <td><video source src="https://github.com/user-attachments/assets/c12c7ff5-ea5a-4e0a-80f9-d984deb8d273"  width="300" height="auto"/> </td>
    <td><video source src="https://github.com/user-attachments/assets/0ab28407-7e57-404c-8e26-77cc256a7ee3"  width="300" height="auto"/></td>
    <td><video source src="https://github.com/user-attachments/assets/706ec048-52cc-469c-b7cd-d7c84aa330b4"  width="300" height="auto"/> </td>
  </tr>
</table> 
</details>

**Comment by Zirochkabila** [comment]
`OsmAnd~ 5.2.0#6052m, Released: 2025-09-15`

Enable pinch zoom magnification - Result

https://github.com/user-attachments/assets/ba42dc85-7fa9-4b03-8a25-54dac1dadd60

**Comment by Zirochkabila** [comment]
`OsmAnd~ 5.2.0#6058m, Released: 2025-09-16`

### Retest

- [x] Remove `Independent features` section
- [x] Move the `Pinch zoom magnification` setting under `Speech Rate`

<details><summary>Details</summary>
<img width="400" alt="Screenshot 2023-01-31 at 11 19 41" src="https://github.com/user-attachments/assets/0647d2df-ad18-4a14-acea-f186f428a4ba">
</details>

**Comment by ihoryakovenko** [comment]
@sonora Android team noticed some bugs https://github.com/osmandapp/OsmAnd/commit/2cc81bd0651d24ea62d27a2c54e5dfaae7ab2665

https://github.com/user-attachments/assets/fa947240-7ac8-40ea-b170-663577ab91e4

**Comment by ihoryakovenko** [comment]
https://github.com/osmandapp/OsmAnd/pull/23489

**Comment by Zirochkabila** [comment]
# Testing 2

`OsmAnd~ 5.2.0#6063m, Released: 2025-09-16`

Tested in both 2D and 3D - **no map jerks observed**
Pinch zoom magnification setting - **moved**



<details><summary>Details</summary>
<table style="width:100%">
  <tr>
    <td><video source src="https://github.com/user-attachments/assets/80eb3e67-d156-4122-a6ce-a5caaf087e4e"  width="300" height="auto"/></td>
<th><img src="https://github.com/user-attachments/assets/4de1be3f-04cf-4879-9b8e-59dddbb63cc1"  width="300" height="auto"/> </th>
  </tr>
</table> 
</details>

**Comment by sonora** [comment]
@Zirochkabila It's only a small quality issue, but I wonder if you can reproduce (point 1 in https://github.com/osmandapp/OsmAnd/issues/15034#issuecomment-3291669599 above):

- Use the V2 engine, with pinch zoom magnification enabled:
- Pinch zoom in some, then leave the fingers static on the screen for a moment and watch the map viewport displayed, or some prominent point on the map.
- Now lift the fingers, and (with the map re-rendering now) observe a small viewport "jump" (the map jumps a little in one direction). The direction and amount of the shift varies, perhaps with where you happen to be on a tile. The effect is rather small in 2D, but can be more apparent in 3D.

There seems a small anchor mismatch between releasing the temporary viewport scaling and the successive map re-render commit. I've been trying to debug it but failed so far. Can you reproduce?

**Comment by Zirochkabila** [comment]
@sonora Yes, it can be reproduced. But it is unclear whether this is a bug or a behavior caused by the fingers making a small movement on the map or not.




<details><summary>Details</summary>
<table style="width:100%">
  <tr>
    <th>On</th>
    <th>Off</th>
  </tr>
  <tr>
    <td><video source src="https://github.com/user-attachments/assets/ec28cdd4-9463-4074-9b0f-8af927f788ef"  width="300" height="auto"/></td>
    <td><video source src="https://github.com/user-attachments/assets/ae827786-9942-44a0-88e2-a6853e167312"  width="300" height="auto"/> </td>
  </tr>
</table> 
</details>

**Comment by sonora** [comment]
Thanks, @Zirochkabila ! I am rather convinced it is a small remaining glitch in the code.

This issue here was about the introduction of the feature, which was a GREAT SUCCESS !. Thanks again to everyone involved, @vshcherb  and especially also @Andrewschenko and @ihoryakovenko .

I have opened a separate #23491 with my interim findings so we can investigate when we get a chance.

**Comment by vshcherb** [comment]
Last commit https://github.com/osmandapp/OsmAnd/commit/2cc81bd0651d24ea62d27a2c54e5dfaae7ab2665 was reverted cause introduce a serious glitch for non-center location (I think without feature enabled). So I think this code is fairly compicated and requires thorough review.


- [ ] To do:  review commit s  Pull request - https://github.com/osmandapp/OsmAnd/pull/23501

**Comment by vshcherb** [comment]
@sonora To be honest for me it looks https://github.com/osmandapp/OsmAnd/commits/master/OsmAnd/src/net/osmand/plus/views/OsmandMapTileView.java too many commits without review in very sensitive area. This is not even about map view 2d, 3d, center positions also Android Auto, Context menus, ... I feel it will be right if revert them all and combine into single pull request (recombine commits probably per feature) for testing and review purposes. Also I feel this review needs to be done by @ihoryakovenko and @alex-dev-neo  to avoid unnecessary later surprises

**Comment by sonora** [comment]
You are right, we probably need a better fix because this one may cause a reference shift if the pinch zoom magnifier is used in motion. Should probably incubate this a bit further.

**Comment by vshcherb** [comment]
- [ ] Zoom in by factor 2-3 produces quite a big shift I think should be fixed

**Comment by ihoryakovenko** [comment]
https://github.com/osmandapp/OsmAnd/pull/23513

**Comment by isso-git** [comment]
Hello,
I am just here to say thank you for implement this great feature also for the rendering version 2.
Apparently, it was a lot of work to implement this.

But in my opinion it was worth it, as I use it so much and it works really well.
I hope many user will find it as part of the Accessibility setting, cause it is very usefull and an unique selling point for Osmand.
As far as I know no other map applications have this feature.

Greetings, thank you very much for your work, please keep up the good work.

**Comment by Gregor0000** [comment]
Hello,
yes I absolutely agree: It's a fantastic feature, I love it. Thank you so very much!
There is only one thing that's broken for me: Pinch & Zoom doesn't work while in the edit mode of the plan route feature.
Please keep up the great work and I hope that you'll find a possibility to fix that 👍

**Comment by DmitryAlexei** [comment]
@Gregor0000 please add steps to reproduce your issue. A short video would  also be very helpful for reproducing.

**Comment by Gregor0000** [comment]
> [@Gregor0000](https://github.com/Gregor0000) please add steps to reproduce your issue. A short video would also be very helpful for reproducing.

I created a screen recording, but I get an error message when I try to upload it here.
However here are the steps to reproduce:
Menu --> Plan route --> Create new route --> Insert a route point --> "Pinch & Zoom" to zoom in --> Issue: Zoom jumps back to previous level when lifting my fingers (so no zoom possible) and center point is shifted/ offsetted afterwards (can only be restored to normal functionality by restarting the app).

**Comment by isso-git** [comment]
Could you please reopen, because I can reproduce the issue of @Gregor0000 above - also with Osmand 5.3.
Steps to reproduce:

- **Activate "Pinch zoom magnification"** in the accessibility plugin
- Start plan a route
- Add ad least one point
- Try to zoom in or out with pinch to zoom

**Actual result:**
It is not possible to zoom by pinch to zoom gesture.
If you use the "+/- Button" on the map zooming work as expected.

**Expected Result:**
Zoom in and out is also possible by pinch to zoom gesture.

Please notice, the error also happen in the edit track function.
Thank you very much for your attention and also for Osmand.

**Comment by Zirochkabila** [comment]
@yuriiurshuliak Please run the test again - there may be a regression

**Comment by sonora** [comment]
Also: It's a great,non-intrusive feature which we hide deep in the Accessibility plugin, and by default it's off. I suggest to at least turn it on by default?

**Comment by Gregor0000** [comment]
@sonora Please don't turn it on by default, unless all problems have been resolved.
I received Osmand 5.3 and this feature is still unusable in combination with "Plan a route".
The "Plan a route" functionality is essential and I use it nearly everyday. That's why I have to keep the new "Pinch & Zoom" feature deactivated in the accessibility plugin. That's unfortunate, because I like this feature a lot. But it has to work flawlessly with "Plan a route" first.

**Comment by sonora** [comment]
@Gregor0000 I suggest to open a new Issue about any findings you still have, as this one here has already been closed.
